aten/src/ATen/native/cuda/SegmentReduce.cu

```diff
@@ -0,0 +1,124 @@
+    return at::_isnan(b) ? b : std::max<OutputT>(a, b);
+  }
+};
+
+Tensor _get_complete_sum(const Tensor& lengths) {
+  int64_t segment_count = lengths.numel();
+  auto offsets = at::empty({segment_count + 1}, lengths.options());
+  offsets[0].zero_();
+  auto* lengths_data_ptr = lengths.data_ptr<int64_t>();
+  auto* offsets_data_ptr = offsets.data_ptr<int64_t>();
+  size_t temp_storage_bytes = 0;
+  cudaError_t error = cub::DeviceScan::InclusiveSum(
+      nullptr,
+      temp_storage_bytes,
+      lengths_data_ptr,
+      offsets_data_ptr + 1,
+      segment_count,
+      at::cuda::getCurrentCUDAStream());
+  TORCH_INTERNAL_ASSERT(error == cudaSuccess, "Failed to make cub call!");
+
+  auto offset_temp_storage = at::empty(
+      {static_cast<int64_t>(temp_storage_bytes)},
+      lengths.options().dtype(kByte));
+
+  error = cub::DeviceScan::InclusiveSum(
+      offset_temp_storage.data_ptr(),
+      temp_storage_bytes,
+      lengths_data_ptr,
+      offsets_data_ptr + 1,
+      segment_count,
+      at::cuda::getCurrentCUDAStream());
+  TORCH_INTERNAL_ASSERT(error == cudaSuccess, "Failed to make cub call!");
+
+  return offsets;
+}
+
+Tensor _segment_reduce_cuda_kernel(
+    const Tensor& data,
+    const c10::optional<Tensor>& lengths,
+    int64_t axis,
+    bool unsafe) {
+  const auto& lengths_value = lengths.value();
+  if (!unsafe) {
+    TORCH_CHECK(
+        (lengths_value >= 0).all().item<int64_t>(),
+        "lengths can not be negative");
+    TORCH_CHECK(lengths_value.sum().item<int64_t>() == data.numel());
+  }
+
+  int64_t segment_count = lengths_value.numel();
+  const auto data_contig = data.contiguous();
+  auto output = at::empty({segment_count}, data.options());
+
+  const auto lengths_contig = lengths_value.contiguous();
+
+  AT_DISPATCH_ALL_TYPES_AND2(
+      at::ScalarType::Half,
+      at::ScalarType::BFloat16,
+      data.scalar_type(),
+      "segment_reduce_cuda",
+      [&]() {
+        auto offsets = _get_complete_sum(lengths_contig);
+        auto* offsets_data_ptr = offsets.data_ptr<int64_t>();
+
+        auto* data_contig_data_ptr = data_contig.data_ptr<scalar_t>();
+        auto* output_data_ptr = output.data_ptr<scalar_t>();
+
+        CustomMax max_op{};
+        size_t temp_storage_bytes = 0;
+        scalar_t initial_value = std::numeric_limits<scalar_t>::lowest();
+        cudaError_t error = cub::DeviceSegmentedReduce::Reduce(
+            nullptr,
+            temp_storage_bytes,
+            data_contig_data_ptr,
+            output_data_ptr,
+            segment_count,
+            offsets_data_ptr,
+            offsets_data_ptr + 1,
+            max_op,
+            initial_value,
+            at::cuda::getCurrentCUDAStream());
+        TORCH_INTERNAL_ASSERT(error == cudaSuccess, "Failed to make cub call!");
+
+        auto temp_storage = at::empty(
+            {static_cast<int64_t>(temp_storage_bytes)},
+            data.options().dtype(kByte));
+        error = cub::DeviceSegmentedReduce::Reduce(
```

Comment:
Please wrap all cub calls in AT_CUDA_CHECK instead of later internal assert (TORCH_INTERNAL_ASSERT means "it's a bug in pytorch", whereas here an error can mean one of many different things wrong with the gpu state):
```
AT_CUDA_CHECK(cub::DeviceSegmentedReduce::Reduce...
);
```